src/module.c

```diff
@@ -60,6 +71,54 @@ int _RegisterDataTypes(RedisModuleCtx *ctx) {
 	return REDISMODULE_OK;
 }
 
+static void _PrepareModuleGlobals() {
+	assert(pthread_mutex_init(&_module_mutex, NULL) == 0);
+	graphs_in_keyspace = array_new(GraphContext *, 1);
+	process_is_child = false;
+}
+
+static void RG_ForkPrepare() {
+	/* At this point, a fork call has been issued. (We assume that this is because BGSave was called.)
+	 * Acquire the writer mutex of each graph to ensure that no locks are held, or else
+	 * the child process will deadlock when attempting to acquire that lock.
+	 * 1. If a writer thread is active, we'll wait until the writer finishes and releases the lock.
+	 * 2. Otherwise, no write in progress. Acquire the lock and release it immediately after forking. */
+
+	// Acquire the module-scoped lock; it will be released after forking.
+	assert(pthread_mutex_lock(&_module_mutex) == 0);
+
+	uint graph_count = array_len(graphs_in_keyspace);
+	for(uint i = 0; i < graph_count; i++) {
+		// Acquire each read-write lock as a reader to guarantee that no graph is being modified.
+		Graph_AcquireReadLock(graphs_in_keyspace[i]->g);
+	}
+}
+
+static void RG_AfterForkParent() {
+	/* The process has forked, and the parent process is continuing.
+	 * Release all locks. */
+
+	uint graph_count = array_len(graphs_in_keyspace);
+	for(uint i = 0; i < graph_count; i++) {
+		// Release each read-write lock.
+		Graph_ReleaseLock(graphs_in_keyspace[i]->g);
+	}
+
+	assert(pthread_mutex_unlock(&_module_mutex) == 0); // Release the module-scoped lock.
+}
+
+static void RG_AfterForkChild() {
+	/* Mark that the child is a forked process so that it doesn't attempt invalid
+	 * accesses of POSIX primitives it doesn't own. */
+	process_is_child = true;
```

Comment:
I assume this logic execute prior to the child dumping logic

Reply:
Correct, this code executes in the child process before any user code does.